addon/wait-for-coroutine.ts

```diff
@@ -0,0 +1,74 @@
+import TestWaiter from './test-waiter';
+
+const COROUTINE_WAITER = new TestWaiter<Generator>('coroutine-waiter');
+
+/**
+ * A convenient utility function to simplify waiting for a coroutine in a
+ * context such as ember-concurrency. Can be used in both decorator and
+ * function form.
+ *
+ * @public
+ * @param coroutine {Generator<T, TReturn, TNext>} The coroutine to track async operations for
+ * @param label {string} An optional string to identify the coroutine
+ *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { waitForGenerator } from 'ember-test-waiters';
+ * import { task } from 'ember-concurrency-decorators';
+ *
+ * export default class Friendz extends Component {
+ *   @task
+ *   @waitForCoroutine
+ *   *doAsyncStuff() {
+ *     yield somethingAsync();
+ *   }
+ * }
+ *
+ * @example
+ *
+ * import Component from '@ember/component';
+ * import { waitForGenerator } from 'ember-test-waiters';
+ * import { task } from 'ember-concurrency';
+ *
+ * export default Component.extend({
+ *   doAsyncStuff: task(waitForCoroutine(function* doAsyncStuff() {
+ *     yield somethingAsync();
+ *   }
+ * });
+ */
+export default function waitForCoroutine(
+  target: object,
+  _key: string,
+  descriptor: PropertyDescriptor,
+  label?: string
+): PropertyDescriptor;
+export default function waitForCoroutine(fn: Function, label?: string): Function;
+
+export default function waitForCoroutine(
+  ...args: [object, string, PropertyDescriptor, string?] | [Function, string?]
+): PropertyDescriptor | Function {
+  if (args.length < 3) {
```

Comment:
I'd prefer if we stored this result in a variable that denoted that this was triggering the functional invocation branch of this code.

```ts
let isFunctionalInvocation = args.length < 3;

if (isFunctionalInvocation) {
  // ...
} else {
  // ...
}
```

Reply:
👍 makes sense, will do

Reply:
Thank you!